pennylane/templates/layers/particle_conserving_u1.py

```diff
@@ -0,0 +1,267 @@
+    :math:`U_A(\phi)` are implemented by the :func:`~.u1_ex_gate` and :func:`~.decompose_ua`
+    functions, respectively.
+
+    Args:
+        weights (array[float]): Array of weights of shape ``(D, M, 2)``.
+            ``D`` is the number of entangler block layers and ``M`` = :math:`N_\mathrm{qubits}-1`
+            is the number of exchange gates :math:`U_{1,\mathrm{ex}}` per layer.
+        wires (Iterable or Wires): Wires that the template acts on. Accepts an iterable of numbers
+            or strings, or a Wires object.
+        init_state (array[int]): Length ``len(wires)`` occupation-number vector representing the
+            HF state. ``init_state`` is used to initialize the wires.
+
+    Raises:
+        ValueError: if inputs do not have the correct format
+
+    .. UsageDetails::
+
+        Notice that:
+
+        #. The number of wires has to be equal to the number of spin orbitals included in
+           the active space.
+
+        #. The number of trainable parameters scales linearly with the number of layers as
+           :math:`2D(N_\mathrm{qubits}-1)`.
+
+        An example of how to use this template is shown below:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.templates import ParticleConservingU1
+            from functools import partial
+
+            # Build the electronic Hamiltonian
+            h, qubits = qml.qchem.molecular_hamiltonian("h2", "h2.xyz")
+
+            # Define the HF state
+            ref_state = qml.qchem.hf_state(electrons=2, qubits)
+
+            # Define the device
+            dev = qml.device('default.qubit', wires=qubits)
+
+            # Define the ansatz
+            ansatz = partial(ParticleConservingU1, init_state=ref_state)
+
+            # Define the cost function
+            cost_fn = qml.VQECost(ansatz, h, dev)
+
+            # Compute the expectation value of 'h'
+            layers = 2
+            params = qml.init.particle_conserving_u1_normal(layers, qubits)
+            print(cost_fn(params))
+    """
+
+    wires = Wires(wires)
+
+    layers = weights.shape[0]
+
+    if len(wires) < 2:
+        raise ValueError(
+            "This template requires the number of qubits to be >= 2; got 'len(wires)' = {}".format(
+                len(wires)
+            )
+        )
+
+    expected_shape = (layers, len(wires) - 1, 2)
+    check_shape(
+        weights,
+        expected_shape,
+        msg="'weights' must be of shape {}; got {}".format(expected_shape, get_shape(weights)),
+    )
+
+    check_type(
+        init_state,
+        [np.ndarray],
+        msg="'init_state' must be a Numpy array; got {}".format(init_state),
+    )
+    for i in init_state:
+        check_type(
+            i,
+            [int, np.int64, np.ndarray],
+            msg="Elements of 'init_state' must be integers; got {}".format(init_state),
+        )
```

Comment:
I was wondering if one should delete this check. From initially checking all sorts of things, we realised that all these checks can be barriers for edge cases. In particular, in tape mode `init_state` _could_ be an interface-type tensor. 

If the type is not wrong, the `BasisState` operation should pick it up and complain.

In summary, I feel that the new guideline for templates is: Only check inputs where you 

* want to look for potentially violated assumptions, like using not enough wires here
* where you need information for further processing, such as the norm of an input state in `AmplitudeEmbedding`
* where there are obvious user mistakes lurking that are hard to understand, such as shapes of weights (which would lead to indexing errors)

Reply:
> * want to look for potentially violated assumptions, like using not enough wires here
> * where you need information for further processing, such as the norm of an input state in AmplitudeEmbedding
> * where there are obvious user mistakes lurking that are hard to understand, such as shapes of weights

Oh, just jumping in to agree, in fact, these are very good guidelines to follow in general 🙂 We should never be trying to validate _everything_; in most cases, we should rely on duck typing and assuming the user is following the docstring.

The main exception would be if there is a combination of input that could cause unexpected issues/errors (such as ints not working where a float might), and would be easy to accidentally pass, especially if calling the function programmatically.

Reply:
This is good news. Thanks.